src/frontend/Deprecation_analysis.ml

```diff
@@ -0,0 +1,168 @@
+open Core_kernel
+open Ast
+open Middle
+
+let deprecated_functions =
+  String.Map.of_alist_exn
+    [ ("multiply_log", "lmultiply")
+    ; ("binomial_coefficient_log", "lchoose")
+    ; ("integrate_ode", "integrate_ode_rk45")
+    ; ("cov_exp_quad", "gp_cov_exp_quad") ]
+
+let deprecated_distributions =
+  String.Map.of_alist_exn
+    (List.concat_map Middle.Stan_math_signatures.distributions
+       ~f:(fun (fnkinds, name, _) ->
+         List.filter_map fnkinds ~f:(function
+           | Lpdf -> Some (name ^ "_log", name ^ "_lpdf")
+           | Lpmf -> Some (name ^ "_log", name ^ "_lpmf")
+           | Cdf -> Some (name ^ "_cdf_log", name ^ "_lcdf")
+           | Ccdf -> Some (name ^ "_ccdf_log", name ^ "_lccdf")
+           | Rng | UnaryVectorized -> None ) ))
+
+let deprecated_userdefined : string String.Table.t = String.Table.create ()
+
+let is_distribution name =
+  Option.is_some (String.Map.find deprecated_distributions name)
+
+let rename_distribution name =
+  Option.value ~default:name (String.Map.find deprecated_distributions name)
+
+let rename_function name =
+  Option.value ~default:name (String.Map.find deprecated_functions name)
+
+let distribution_suffix name =
+  String.is_suffix ~suffix:"_lpdf" name
+  || String.is_suffix ~suffix:"_lpmf" name
+  || String.is_suffix ~suffix:"_lcdf" name
+  || String.is_suffix ~suffix:"_lccdf" name
+
+let userdef_distributions stmts =
+  List.filter_map
+    ~f:(function
+      | {stmt= FunDef {funname= {name; _}; _}; _} ->
+          if
+            String.is_suffix ~suffix:"_log_lpdf" name
+            || String.is_suffix ~suffix:"_log_lpmf" name
+          then Some (String.drop_suffix name 5)
+          else if String.is_suffix ~suffix:"_log_log" name then
+            Some (String.drop_suffix name 4)
+          else None
+      | _ -> None)
+    (Option.value ~default:[] stmts)
+
+let without_suffix user_dists name =
+  if
+    String.is_suffix ~suffix:"_lpdf" name
+    || String.is_suffix ~suffix:"_lpmf" name
+  then String.drop_suffix name 5
+  else if
+    String.is_suffix ~suffix:"_log" name
+    && not
+         ( is_distribution (name ^ "_log")
+         || List.exists ~f:(( = ) name) user_dists )
+  then String.drop_suffix name 4
+  else name
+
+let warn_deprecated (loc_span, message) =
+  Fmt.pf Fmt.stderr
+    "@[<v>@,Warning: deprecated language construct used in %s:@,%s@]@."
+    (Location_span.to_string loc_span)
+    message
+
+let replace_suffix = function
```

Comment:
This looks like it's also back in the old file.

Reply:
Thanks - I tried to comment it out and got an `unbound value` error. I now see that when I add it to `Deprecation_analysis.mli`, it resolves the problem.